cognite/preprocessing.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+'''Preprocessing module.
+
+This module provides a number of preprocessing methods to further clean the data retrieved from the Cognite API.
+'''
+
+import math
+import warnings
+from functools import reduce
+
+import numpy as np
+import pandas as pd
+
+
+def _merge_dataframes(df1, df2):
+    if df1 is None:
+        return df2
+
+    if df2 is None:
+        return df1
+
+    return pd.merge(df1, df2, on='timestamp', how='outer') \
+        .sort_values(by='timestamp') \
+        .reset_index(drop=True)
+
+
+def merge_list_of_dataframes(dataframes):
+    """Merges together a list of dataframes and creates a time index with evenly spaced intervals.
+
+    Adds NaN to timestamps with missing values.
+
+    Args:
+        dataframes (list(pandas.DataFrame)):  Input dataframes.
+
+    Returns:
+        pandas.DataFrame: Dataframes merged into one with evenly spaced intervals.
+    """
+    df_merged = reduce(lambda x, y: _merge_dataframes(x, y), dataframes)
+    df_even_index = make_index_even(df_merged)
+    return df_even_index
+
+
+def make_index_even(dataframe):
+    """Creates time index with evenly spaced intervals. Adds NaN to timestamps with missing values.
+
+    Args:
+        dataframe (pandas.DataFrame):  Input dataframe.
+
+    Returns:
+        pandas.DataFrame: Input dataframe with evenly spaced intervals.
+    """
+    timestamps = dataframe.timestamp.values
+    start_time = timestamps[0]
+    end_time = timestamps[-1]
+    deltas = np.diff(timestamps, 1)
+    delta = reduce(lambda x, y: math.gcd(x, y), deltas)  # gcd of deltas
+    t_new = np.arange(start_time, end_time + delta, delta)
+    new_df = pd.DataFrame(t_new, columns=['timestamp'])
+    return new_df.merge(dataframe, on='timestamp', how='outer').sort_values(by='timestamp').reset_index(drop=True)
+
+
+def fill_nan(dataframe):
+    """Uses step interpolation to replace NaN values with the previous non-NaN value.
+
+    Args:
+        dataframe (pandas.DataFrame):  Input dataframe.
+
+    Returns:
+        pandas.DataFrame: Input dataframe with NaN values removed by forward fill.
+    """
+    return dataframe.fillna(method='ffill')
+
+
+def remove_nan_columns(dataframe):
+    """Removes columns of data frame where any value is NaN.
+
+    Args:
+        dataframe (pandas.DataFrame):  Input dataframe.
+
+    Returns:
+        tuple: tuple containing:
+            pandas.DataFrame: Dataframe with columns containing NaN values removed.
+            numpy.array: Array of bools indicating which columns were kept.
+    """
+    df_copy = dataframe.set_index('timestamp')
+    selected_columns_mask = df_copy.notnull().all().values
+    return dataframe.dropna(axis=1, how='any'), selected_columns_mask
+
+
+def normalize(dataframe):
```

Comment:
This will fail if one of the columns is constant. Something we should let the user know? Only matters if zero-variance columns have not been removed.